### PR TITLE
Implement async_checkpoint

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #302

Summary:
This PR implements 2 different async checkpoint. The first one is to use DCP.async_save another one is to use pinned memory + a seperate process to avoid GILs issue.

### CODE REVIEW

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -290,6 +290,23 @@ def __init__(self):
...
         )
+        self.parser.add_argument(
+            "--checkpoint.async_mode",
+            type=str,
+            default="disabled",
+            help="""
+                Which async checkpoint mode to use. Currently there are 3 different modes.
+                1. "disabled": synchronized checkpointing will be used.
+                2. "async": torch.distributed.checkpoint.async_save will be used.
+                3. "async_with_pinned_mem": a dedicated pinned memory space will be used and
```
Comment (awgu) [inline torchtitan/config_manager.py]: What is the recommendation to users on how to choose between `"async"` vs. `"async_with_pinned_mem"`?

How should users think about the tradeoffs? For example, is the difference only about performance? Is there a system compatibility constraint? (e.g. how much CPU RAM is required?)

Comment (fegin) [reply torchtitan/config_manager.py]: Good point, I'll add the memory constraint to the description Using `async` may be enough for most users. `async_with_pinned_mem` should be used when users pursue zero-overhead (almost zero-overhead) checkpoint. Will also add this to the description.

File: torchtitan/checkpoint.py [hunk 2]
```diff
@@ -160,19 +211,125 @@ def save(self, curr_step: int, force: bool = False) -> None:
...
+                return False
+            if self.interval_type == IntervalType.SECONDS:
+                time_sync_result = (time.monotonic() - self.begin_time) >= self.interval
+                self.time_sync_result = torch.tensor(int(time_sync_result))
+                if self.time_sync_work is None:
+                    self.time_sync_work = dist.all_reduce(
+                        self.time_sync_result, group=self.pg, async_op=True
+                    )
+                    return False
+                elif curr_step % 5 == 4:
```
Comment (lessw2020) [inline torchtitan/checkpoint.py:240]: I guess 5 seconds is a good default for synching, but should this be exposed as a class or var ala _interval_sync_step = 5, and then it becomes 
curr_step % _interval_sync_step = _interval_sync_step -1?

**Review by lessw2020 — APPROVED** [review]
fantastic feature add! 
works great - tested all three types and toggled around with load and save as well.

one general comment - the time.monotonic prints in the logging are not obvious what they are without seeing the code and I think it could be disconcerting to users.
Can it be modified to {time.monotonic=} so that it will print:
~~~
Waiting for the background process to finish, time.monotonic = 1800617.210797273.
~~~
instead of just:
~~~
Waiting for the background process to finish, 1800617.210797273. 
~~~
basically at least provide some info re: these numbers that just start appearing in the logs. 
and possibly add a prefix of [checkpointing] to group these entries, showing they are checkpointing specific entries (also makes the logs more searchable if needed). 
Anyway, great feature add!

File: torchtitan/checkpoint.py (hunk 2 above)
Comment (fegin) [reply torchtitan/checkpoint.py:240]: This time interval was added before, we can change it in a separate PR.
